pkg/binding/crd.go

```diff
@@ -58,69 +97,32 @@ func (c *Controller) handleCRD(_ runtime.Object) error {
-func (c *Controller) checkAPIResourcesForUpdates() ([]APIResource, []string, error) {
-	toStart := []APIResource{}
-	toStop := []string{}
-
-	// tracking keys are used to detect what API resources have been removed
-	trackingKeys := map[string]bool{}
-	for k := range c.informers {
-		trackingKeys[k] = true
+func (c *Controller) includedToWatch(r APIResource) bool {
+	if _, excluded := excludedGroups[r.groupVersion.Group]; excluded {
+		return false
 	}
-
-	// Get all the api resources in the cluster
-	apiResources, err := c.kubernetesClient.Discovery().ServerPreferredResources()
-	if err != nil {
-		// ignore the error caused by a stale API service
-		if !strings.Contains(err.Error(), util.UnableToRetrieveCompleteAPIListError) {
-			return nil, nil, err
-		}
+	if !util.IsAPIGroupAllowed(r.groupVersion.Group, c.allowedGroupsSet) {
+		return false
 	}
-
-	// Loop through the api resources and create informers and listers for each of them
-	for _, list := range apiResources {
-		gv, err := schema.ParseGroupVersion(list.GroupVersion)
-		if err != nil {
-			c.logger.Error(err, "Failed to parse a GroupVersion", "groupVersion", list.GroupVersion)
-			continue
-		}
-		if _, excluded := excludedGroups[gv.Group]; excluded {
-			continue
-		}
-		for _, resource := range list.APIResources {
-			if _, excluded := excludedResourceNames[resource.Name]; excluded {
-				continue
-			}
-			if !util.IsAPIGroupAllowed(gv.Group, c.allowedGroupsSet) {
-				continue
-			}
-			informable := verbsSupportInformers(resource.Verbs)
-			if informable {
-				key := util.KeyForGroupVersionKind(gv.Group, gv.Version, resource.Kind)
-				if _, ok := c.informers[key]; !ok {
-					toStart = append(toStart, APIResource{
-						groupVersion: gv,
-						resource:     resource,
-					})
-				}
-				// remove the key from tracking keys, what is left in the map at the end are
-				// keys to the informers that need to be stopped.
-				delete(trackingKeys, key)
-			}
-		}
+	if _, excluded := excludedResourceNames[r.resource.Name]; excluded {
+		return false
 	}
+	return true
+}
 
-	for k := range trackingKeys {
-		toStop = append(toStop, k)
+func crdEstablished(crd *apiextensionsv1.CustomResourceDefinition) bool {
+	for _, condition := range crd.Status.Conditions {
+		if condition.Type == apiextensionsv1.Established && condition.Status == apiextensionsv1.ConditionTrue {
+			return true
+		}
 	}
-	return toStart, toStop, nil
+	return false
 }
 
-func (c *Controller) startInformersForNewAPIResources(toStartList []APIResource) {
+func (c *Controller) startInformersForNewAPIResources(ctx context.Context, toStartList []APIResource) {
+	logger := klog.FromContext(ctx)
 	for _, toStart := range toStartList {
-		c.logger.Info("New API added. Starting informer for:", "group", toStart.groupVersion.Group,
+		logger.Info("Ensuring informer for:", "group", toStart.groupVersion.Group,
 			"version", toStart.groupVersion, "kind", toStart.resource.Kind)
 
```

Comment:
I see a lot of work below that could be conditional on whether `key` is already in `c.informers`.

Reply:
Done in latest push.

Reply:
I see the early-out now. I am surprised by the retention of the complex code later for updating `c.listers`, `c.informers`, and `c.stoppers`. Aren't they coordinated? For a given key, either all three have an entry or all three do not have an entry, right? BTW, this would be more clearly handled by one map from key to a pointer to a struct holding the informer, lister, and stopper.